src/backend/cuda/nvrtc/launch_args.hpp

```diff
@@ -0,0 +1,137 @@
+/*******************************************************
+ * Copyright (c) 2019, ArrayFire
+ * All rights reserved.
+ *
+ * This file is distributed under 3-clause BSD license.
+ * The complete license agreement can be obtained at:
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+
+#pragma once
+
+#include <cuda.h>
+#include <cuda_runtime.h>
+#include <nvrtc.h>
+
+#include <vector>
+
+namespace cuda {
+
+/**
+    EnqueueArgs is a kernel launch configuration composition object
+
+    This structure is an composition of various parameters that are
+    required to successfully launch a CUDA kernel.
+ */
+struct EnqueueArgs {
+//TODO(pradeep): this can be easily templated
+//template<typename Queue, typename Event>
+    CUstream mStream;               ///< CUDA stream to enqueue the kernel on
+    dim3 mBlocks;                   ///< Number of blocks per grid/kernel-launch
+    dim3 mThreads;                  ///< Number of threads per block
+    unsigned int mSharedMemSize;    ///< Size(in bytes) of shared memory used
+    std::vector<CUevent> mEvents;   ///< Events to wait for kernel execution
+
+    /*
+        \brief EnqueueArgs constructor
+
+        \param[in] blks is number of blocks per grid
+        \param[in] thrds is number of threads per block
+        \param[in] sharedMemSize is number of bytes of shared memory allocation
+    */
+    EnqueueArgs(dim3 blks, dim3 thrds,
+            unsigned int sharedMemSize=0) :
+      mStream(0),
+      mBlocks(blks),
+      mThreads(thrds),
+      mSharedMemSize(sharedMemSize) {
+    }
+
+    /*
+        \brief EnqueueArgs constructor
+
+        \param[in] event is an event for which kernel execution has to wait for
+        \param[in] blks is number of blocks per grid
+        \param[in] thrds is number of threads per block
+        \param[in] sharedMemSize is number of bytes of shared memory allocation
+    */
+    EnqueueArgs(CUevent event, dim3 blks, dim3 thrds,
```

Comment:
Blocks and threads should be first always. The optional stuff should be at the end.